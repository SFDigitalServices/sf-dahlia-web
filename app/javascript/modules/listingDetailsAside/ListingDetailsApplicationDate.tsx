import React from "react"
import { RailsListing } from "../listings/SharedHelpers"
import { ApplicationStatus, ApplicationStatusType, Icon, t } from "@bloom-housing/ui-components"
import { localizedFormat } from "../../util/languageUtil"
import dayjs from "dayjs"
import { LISTING_TYPE_FIRST_COME_FIRST_SERVED } from "../constants"
import { useFeatureFlag } from "../../hooks/useFeatureFlag"
import { isFcfsSalesListing, isOpen } from "../../util/listingUtil"
import bloomTheme from "../../../../tailwind.config"
import { Message } from "@bloom-housing/ui-seeds"

export interface ListingDetailsApplicationDateProps {
  listing: RailsListing
}

export const formatTime = (time: string) => {
  const formattedTime = dayjs(time).format("h:mm")
  const hour = Number(dayjs(time).format("H"))
  // \u00A0 is a non-breaking space
  const suffix = hour >= 12 ? "\u00A0PM" : "\u00A0AM"
  return `${formattedTime}${suffix}`
}

const StatusMessage = ({ isClosed, message }: { isClosed: boolean; message: string }) => {
  return (
    <Message
      fullwidth
      className="justify-start leading-5"
      variant={isClosed ? "alert" : "primary"}
      customIcon={<Icon fill={isClosed ? "red-700" : ""} symbol="clock" size="medium" />}
    >
      {message}
    </Message>
  )
}

const getStatusLottery = (listing: RailsListing) => {
  const isClosed = !isOpen(listing)
  const datetime = listing.Application_Due_Date
  return (
    <StatusMessage
      isClosed={isClosed}
      message={t(
        isClosed
          ? "listingDetails.applicationsClosed.withDateTime"
          : "listingDetails.applicationsDeadline.withDateTime",
        {
          date: localizedFormat(datetime, "LL"),
          time: formatTime(datetime),
        }
      )}
    />
  )
}

const getStatusFcfs = (listing: RailsListing) => {
  const isClosed = !listing.Accepting_Online_Applications
  const datetime = listing.Application_Start_Date_Time
  const showDateTime = new Date(listing.Application_Start_Date_Time) > new Date()

  const getMessage = () => {
    if (isClosed) {
      return t("listingDetails.applicationsClosed")
    } else {
      return showDateTime
        ? t("listingDetails.applicationsOpen.withDateTime", {
            date: localizedFormat(datetime, "LL"),
            time: formatTime(datetime),
          })
        : t("listingDetails.applicationsOpen")
    }
  }

  return <StatusMessage isClosed={isClosed} message={getMessage()} />
}

const ListingDetailsStatus = ({ listing }: { listing: RailsListing }) => {
  return listing.Listing_Type === LISTING_TYPE_FIRST_COME_FIRST_SERVED
    ? getStatusFcfs(listing)
    : getStatusLottery(listing)
}

export const ListingDetailsApplicationDate = ({ listing }: ListingDetailsApplicationDateProps) => {
  const isApplicationOpen = listing && isOpen(listing)

  const { unleashFlag: isSalesFcfsEnabled } = useFeatureFlag("FCFS", false)

  return (
    <div className="w-full mb-8 md:mb-0">
      {isSalesFcfsEnabled && isFcfsSalesListing(listing) ? (
        <ListingDetailsStatus listing={listing} />
      ) : (
        <div className="w-full mb-8 md:mb-0">
          <ApplicationStatus
            className="place-content-center"
            content={t(
              isApplicationOpen
                ? "listingDetails.applicationDeadline.open"
                : "listingDetails.applicationDeadline.closed",
              {
                date: localizedFormat(listing.Application_Due_Date, "ll"),
                time: dayjs(listing.Application_Due_Date).format("h:mm A"),
              }
            )}
            iconColor={!isApplicationOpen && bloomTheme.theme.colors.red["700"]}
            status={isApplicationOpen ? ApplicationStatusType.Open : ApplicationStatusType.Closed}
          />
        </div>
      )}
    </div>
  )
}
