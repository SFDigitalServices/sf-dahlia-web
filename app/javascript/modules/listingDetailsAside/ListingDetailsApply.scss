.fcfs-bmr-how-to-apply {
  padding: var(--seeds-spacer-content) !important;
}

.fcfs-bmr-how-to-apply__list {
  padding-top: var(--seeds-spacer-content) !important;
  padding-bottom: var(--seeds-spacer-content) !important;
}

.fcfs-bmr-how-to-apply {
  h4.text__underline-weighted {
    margin-bottom: 0px;
  }
}
