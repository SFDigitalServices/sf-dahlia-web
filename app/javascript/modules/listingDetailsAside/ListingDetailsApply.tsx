import React, { useState } from "react"
import {
  AppearanceStyleType,
  Button,
  Heading,
  LinkButton,
  MultiLineAddress,
  OrDivider,
  SidebarBlock,
  t,
} from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import {
  acceptingPaperApplications,
  isHabitatListing,
  isOpen,
  isRental,
  isSale,
  paperApplicationURLs,
} from "../../util/listingUtil"
import { localizedFormat, renderInlineMarkup } from "../../util/languageUtil"

export interface ListingDetailsApplyProps {
  listing: RailsListing
}

export const ListingDetailsApply = ({ listing }: ListingDetailsApplyProps) => {
  const [paperApplicationsOpen, setPaperApplicationsOpen] = useState(false)

  if (!isOpen(listing)) return null

  const isListingRental = isRental(listing)

  const acceptingPaperApps = acceptingPaperApplications(listing)

  const ordinalHeader = (ordinal: number, title: string) => {
    return (
      <Heading priority={4} className={"text-gray-900 text-lg -mt-2 mb-3"}>
        <span className={"text-blue-600 mr-2"}>{ordinal}</span>
        {title}
      </Heading>
    )
  }

  const howToApplyBlock = (
    <SidebarBlock title={t("listings.apply.howToApply")}>
      {!isListingRental && (
        <>
          <p className={"mb-4"}>
            {renderInlineMarkup(
              t("listings.apply.fulfillEligibilityRequirements", {
                url: isHabitatListing(listing)
                  ? "https://habitatgsf.org/amber-drive-info/"
                  : "https://sfmohcd.org/homebuyer-program-eligibility",
              })
            )}
          </p>
          <p className={"mb-4"}>{t("listings.apply.eligibilityRequirementDescription")}</p>
        </>
      )}
      <LinkButton
        styleType={AppearanceStyleType.primary}
        className={"w-full"}
        transition={true}
        href={`listings/${listing.listingID}/apply-welcome/intro`}
      >
        {t("label.applyOnline")}
      </LinkButton>
      {process.env.COVID_UPDATE && (
        <div className={"mt-4"}>
          <Heading priority={4} className={"text-base text-gray-800 font-sans"}>
            {t("listings.apply.covidUpdate")}
          </Heading>
          <div className={"text-gray-700 text-base mt-2"}>
            {t("listings.apply.covidUpdateInfo")}
          </div>
        </div>
      )}
      {acceptingPaperApps && (
        <>
          <OrDivider bgColor={"white"} />
          {ordinalHeader(1, t("listings.apply.getAPaperApplication"))}
          {t("listings.apply.paperApplicationsMustBeMailed")}
          <Button
            className={"w-full mt-4"}
            transition={true}
            icon={"arrowDown"}
            iconPlacement={"right"}
            onClick={() => setPaperApplicationsOpen(!paperApplicationsOpen)}
          >
            {t("label.downloadApplication")}
          </Button>
          {paperApplicationsOpen && acceptingPaperApplications && (
            <div className={"flex w-full items-center justify-center flex-col"}>
              {paperApplicationURLs(isListingRental).map((app) => {
                return (
                  <span className={"pt-4"}>
                    <LinkButton
                      href={app.fileURL}
                      unstyled
                      className={"m-0 no-underline"}
                      newTab={true}
                    >
                      {app.languageString}
                    </LinkButton>
                  </span>
                )
              })}
            </div>
          )}
        </>
      )}
    </SidebarBlock>
  )

  const submitPaperApplicationBlocks = (
    <>
      <SidebarBlock className={"bg-blue-200"}>
        {ordinalHeader(2, t("listings.apply.submitAPaperApplication"))}
        {t("listings.apply.includeAnEnvelope")}
      </SidebarBlock>
      <SidebarBlock
        className={"bg-blue-200"}
        styleType={"capsWeighted"}
        title={t("listings.apply.sendByUsMail")}
      >
        <div className={"mb-2 text-gray-900 text-base"}>
          <MultiLineAddress
            address={{
              placeName: listing.Application_Organization,
              street: listing.Application_Street_Address,
              city: listing.Application_City,
              zipCode: listing.Application_Postal_Code,
              state: listing.Application_State,
            }}
          />
        </div>
        {t("listings.apply.applicationsMustBeReceivedByDeadline")}
      </SidebarBlock>
      <SidebarBlock className={"bg-blue-200"}>
        {t("listings.doNotApplyOneAndPaperOrMultiple")}
      </SidebarBlock>
    </>
  )

  const needHelpBlock = (
    <SidebarBlock title={t("listings.apply.needHelp")}>
      {isListingRental && (
        <div className={"mb-4"}>{t("listings.apply.visitAHousingCounselor")}</div>
      )}
      <LinkButton
        transition={true}
        newTab={true}
        href={
          !isListingRental
            ? "https://www.homeownershipsf.org/application-assistance-for-homebuyers/"
            : "/housing-counselors"
        }
        className={"w-full"}
      >
        {isListingRental
          ? t("housingCounselor.findAHousingCounselor")
          : t("listings.apply.visitHomeownershipSf")}
      </LinkButton>
    </SidebarBlock>
  )

  const expectedMoveInDateBlock = (
    <SidebarBlock title={t("listings.expectedMoveinDate")}>
      {localizedFormat(listing.Expected_Move_in_Date, "MMMM YYYY")}
    </SidebarBlock>
  )

  return (
    <>
      {howToApplyBlock}
      {acceptingPaperApps && submitPaperApplicationBlocks}
      {needHelpBlock}
      {isSale(listing) && listing.Expected_Move_in_Date && expectedMoveInDateBlock}
    </>
  )
}
