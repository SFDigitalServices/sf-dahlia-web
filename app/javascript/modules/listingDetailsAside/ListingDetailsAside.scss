.fcfs-aside {
    --card-border-radius: rounded-none;
    --card-background-color: var(--bloom-color-gray-200);
    --card-footer-background-color: var(--bloom-color-gray-200);
    --card-border-width: border-b-1;

    .aside-card-header {
        font-family: var(--seeds-font-alt-sans);
        font-size: var(--seeds-font-size-lg);
        font-weight: var(-seeds-font-weight-semibold);
        line-height: var(--seeds-line-height-6);
    }

    .aside-card-section {
        font-family: var(--seeds-font-sans);
        font-size: var(--seeds-font-size-sm);
        font-weight: var(--seeds-font-weight-regular);
        line-height: var(--seeds-line-height-5);
    }
}
