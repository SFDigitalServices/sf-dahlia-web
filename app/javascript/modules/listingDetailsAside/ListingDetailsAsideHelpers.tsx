import { t } from "@bloom-housing/ui-components"
import React from "react"

export const SeeThisUnitMLS = ({ url }: { url: string }) => {
  return (
    <p className="mt-1">
      <a href={url} target="_blank">
        {t("listings.process.seeThisUnitOnMls")}
      </a>
    </p>
  )
}
