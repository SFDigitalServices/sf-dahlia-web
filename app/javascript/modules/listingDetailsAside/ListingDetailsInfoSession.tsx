import React from "react"
import { getEventNote, RailsListing } from "../listings/SharedHelpers"
import { EventSection, t } from "@bloom-housing/ui-components"
import { localizedFormat } from "../../util/languageUtil"
import { getEventTimeString } from "../../util/listingUtil"

export interface ListingDetailsInfoSessionProps {
  listing: RailsListing
}

export const ListingDetailsInfoSession = ({ listing }: ListingDetailsInfoSessionProps) => {
  return (
    <>
      {listing.Information_Sessions?.length > 0 ? (
        <EventSection
          dateClassName="normal-case"
          sectionHeader={true}
          events={listing.Information_Sessions?.map((informationSession) => {
            return {
              dateString: informationSession.Date && localizedFormat(informationSession.Date, "LL"),
              timeString: getEventTimeString(informationSession),
              note: getEventNote(informationSession),
            }
          })}
          headerText={t("listings.process.informationSessions")}
        />
      ) : null}
      {/* TODO: Bloom prop changes for get and submit application sections */}
    </>
  )
}
