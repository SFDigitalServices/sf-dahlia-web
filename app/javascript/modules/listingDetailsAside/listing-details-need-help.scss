.listing-details-need-help {
  padding: var(--seeds-spacer-content) !important;
  border-color: var(--bloom-color-gray-400) !important;
}
