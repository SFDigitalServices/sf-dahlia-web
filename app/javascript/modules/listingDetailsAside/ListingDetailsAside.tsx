import React from "react"
import { ListingDetailItem, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsInfoSession } from "./ListingDetailsInfoSession"
import { ListingDetailsProcess } from "./ListingDetailsProcess"
import { isOpen } from "../../util/listingUtil"
import { ListingDetailsApply } from "./ListingDetailsApply"
import { ListingDetailsApplicationDate } from "./ListingDetailsApplicationDate"
import { ListingDetailsLotteryResults } from "../listingDetailsLottery/ListingDetailsLotteryResults"

export interface ListingDetailsSidebarProps {
  listing: RailsListing
  imageSrc: string
}

export const ListingDetailsAside = ({ listing, imageSrc }: ListingDetailsSidebarProps) => {
  const isApplicationOpen = isOpen(listing)

  return (
    <ListingDetailItem
      imageAlt={""}
      imageSrc={imageSrc}
      title={t("listings.process.header")}
      subtitle={t("listings.process.subheader")}
      hideHeader={true}
      desktopClass="header-hidden"
    >
      <aside className="w-full static md:absolute md:right-0 md:w-1/3 md:top-0 sm:w-2/3 md:ml-2 h-full md:border border-solid bg-white">
        <div className="hidden md:block">
          <ListingDetailsApplicationDate isApplicationOpen={isApplicationOpen} listing={listing} />
          <ListingDetailsLotteryResults listing={listing} />
          {isApplicationOpen && <ListingDetailsInfoSession listing={listing} />}
          <ListingDetailsApply listing={listing} />
          <ListingDetailsProcess listing={listing} />
        </div>
      </aside>
    </ListingDetailItem>
  )
}
