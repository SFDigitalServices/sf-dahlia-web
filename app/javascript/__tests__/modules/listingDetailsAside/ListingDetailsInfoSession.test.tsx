import React from "react"
import renderer from "react-test-renderer"
import { ListingDetailsInfoSession } from "../../../modules/listingDetailsAside/ListingDetailsInfoSession"
import { openSaleListing } from "../../data/RailsSaleListing/listing-sale-open"
import { closedRentalListing } from "../../data/RailsRentalListing/listing-rental-closed"

describe("ListingDetailsInfoSession", () => {
  it("renders ListingDetailsInfoSession component", () => {
    const tree = renderer.create(<ListingDetailsInfoSession listing={openSaleListing} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("does not render ListingDetailsInfoSession component whith no information sessions attached to listing", () => {
    const closedListingWithoutInfoSessions = {
      ...closedRentalListing,
      Information_Sessions: [],
    }

    const tree = renderer
      .create(<ListingDetailsInfoSession listing={closedListingWithoutInfoSessions} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
