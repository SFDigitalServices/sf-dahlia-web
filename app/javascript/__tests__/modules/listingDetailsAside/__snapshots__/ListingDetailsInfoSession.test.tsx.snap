// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsInfoSession does not render ListingDetailsInfoSession component whith no information sessions attached to listing 1`] = `<DocumentFragment />`;

exports[`ListingDetailsInfoSession renders ListingDetailsInfoSession component 1`] = `
<DocumentFragment>
  <div
    class="border-b border-gray-400 md:border-b-0"
  >
    <section
      class="aside-block"
    >
      <h4
        class="text__underline-weighted"
      >
        Information Sessions
      </h4>
      <div
        class="false"
      >
        <p
          class="text text-gray-800 pb-2 flex justify-between items-center"
        >
          <span
            class="inline-block text-sm uppercase normal-case"
          >
            November 18, 2019
          </span>
          <span
            class="inline-block text-xs font-bold ml-5 font-alt-sans"
          >
            6:00 PM
          </span>
        </p>
        <div
          class="text-sm text-gray-700 false"
        >
          <div
            class="flex flex-col"
          >
            <span
              class="links-space translate"
            >
              <span>
                MOHCD
              </span>
            </span>
            <span>
              1 South Van Ness Ave, Fl 5, San Francisco
            </span>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
